---
title: "WES"
author: "yinghong"
date: "2024-07-31"
output: html_document
---

# load packages
```{r}
library(maftools)
library(tidyverse)
library(stringr)
library(tidyr)
library(circlize)
library(ComplexHeatmap)
library(data.table)
library(dplyr)
library(stringr)
library(tidyr)
library(pheatmap)
library(ComplexHeatmap)
library(ggplot2)
library(reshape2)
library(pvclust)
library(ape)
library(phangorn)
library(ggpubr)
library(FSA)
library(readr)

# dir.create("../01.Data")
# dir.create("../02.Code")
# dir.create("../03.Output")

```

# 1.read maf files
```{r}
rm(list=ls())

file_pattern <- "../01.Data/maf/*.maf"

files <- list.files(path = "../01.Data/maf", pattern = "\\.maf$", full.names = TRUE)

sample_depths <- lapply(files, function(file) {

  maf <- read_tsv(file, col_types = cols(.default = "c"))

  maf <- maf %>%
    mutate(
      Ref_allele_depth = as.numeric(Ref_allele_depth),
      Alt_allele_depth = as.numeric(Alt_allele_depth)
    )
  
  maf <- maf %>%
    mutate(total_depth = Ref_allele_depth + Alt_allele_depth)

  avg_depth <- mean(maf$total_depth, na.rm = TRUE)
  
  sample_name <- strsplit(basename(file), "\\.")[[1]][1]
  
  return(data.frame(Sample = sample_name, Avg_Depth = avg_depth))
}) %>% bind_rows()

overall_avg_depth <- mean(sample_depths$Avg_Depth, na.rm = TRUE)

```

# 2.merge maf files
```{r}

all_samples_maf <- lapply(files, function(file) {

  maf <- read_tsv(file, col_types = cols(.default = "c"))

  sample_name <- strsplit(basename(file), "\\.")[[1]][1]
  maf <- maf %>% mutate(Tumor_Sample_Barcode = sample_name)
  return(maf)
}) %>% bind_rows()

output_file <- "../03.Output/merged_samples_maf.maf"
write_tsv(all_samples_maf, output_file)

maf <- read_tsv(output_file)

maf <- maf %>%
  mutate(total_depth = Ref_allele_depth + Alt_allele_depth)

```

# 3.merge annovar files
```{r}

file_pattern <- "../01.Data/annovar_results/*somatic.annovar.hg19_multianno.txt"

files <- list.files(path = "../01.Data/annovar_results", pattern = "somatic.annovar.hg19_multianno.txt", full.names = TRUE)

all_samples_df <- lapply(files, function(file) {
  sample_name <- strsplit(basename(file), "\\.")[[1]][1]
  df <- read_tsv(file)
  df$Sample <- sample_name
  return(df)
}) %>% bind_rows()

```

# 4.mean sequencing depth
```{r}

duplicated_all_samples_df <- all_samples_df %>%
  group_by(Chr, Start, End, Sample) %>%
  filter(n() > 1)

duplicated_maf <- maf %>%
  group_by(Chromosome, Start_Position, End_Position, Tumor_Sample_Barcode) %>%
  filter(n() > 1)

all_samples_df <- all_samples_df %>%
  distinct(Chr, Start, End, Sample, .keep_all = TRUE)

maf <- maf %>%
  distinct(Chromosome, Start_Position, End_Position, Tumor_Sample_Barcode, .keep_all = TRUE)

maf_summary <- maf %>%
  group_by(Chromosome, Start_Position, End_Position, Tumor_Sample_Barcode) %>%
  summarise(total_depth = mean(total_depth, na.rm = TRUE), .groups = 'drop')

mean(maf_summary$total_depth) 

```

# 5.depth filter 
```{r}
maf$Chromosome <- paste0("chr",maf$Chromosome )

merged_df <- all_samples_df %>%
  left_join(maf %>% select(Chromosome, Start_Position, End_Position, total_depth, Tumor_Sample_Barcode), 
            by = c("Chr" = "Chromosome", "Start" = "Start_Position", "End" = "End_Position", "Sample" = "Tumor_Sample_Barcode"))

merged_df$Tumor_Sample_Barcode <- merged_df$Sample

depth_threshold <- 10
filtered_df <- merged_df %>% filter(total_depth >= depth_threshold)

frequency_threshold <- 0.01
filtered_df <- filtered_df %>%
  filter( gnomAD_exome_ALL <= frequency_threshold)

mutation_counts <- filtered_df%>%
  filter(ExonicFunc.refGene %in% c("nonsynonymous SNV", "synonymous SNV")) %>%
  group_by(ExonicFunc.refGene) %>%
  summarise(count = n())

```

# 6.keep snv
```{r}

functional_variants <- c("nonsynonymous SNV", "nonframeshift substitution", "stopgain", 
                   "nonframeshift deletion", "frameshift deletion", "frameshift insertion", 
                   "stoploss", "frameshift substitution")
filtered_df <- filtered_df %>% filter(ExonicFunc.refGene %in% functional_variants)

filtered_df$Otherinfo <- gsub("\t","",filtered_df$Otherinfo)

filtered_output_file <- "../03.Output/filtered_annovar_output.txt"
write_tsv(filtered_df, filtered_output_file)

variant_counts <- filtered_df %>%
  group_by(Sample) %>%
  summarise(Variant_Count = n())

ggplot(variant_counts, aes(x = Sample, y = Variant_Count)) +
  geom_bar(stat = "identity") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  labs(title = "Variant Counts per Sample", x = "Sample", y = "Number of Variants")

ggsave("../03.Output/variant_counts_per_sample.png")

```


```{r}
library(dplyr)

mutation_counts <- filtered_df  %>%
  group_by(Sample, ExonicFunc.refGene) %>%
  summarise(count = n()) %>%
  ungroup()

variant_counts <- filtered_df %>%
  group_by(Sample) %>%
  summarise(Variant_Count = n())

mutation_counts <- mutation_counts %>%
  arrange(Sample, desc(ExonicFunc.refGene))

mutation_counts <- mutation_counts %>%
  group_by(Sample) %>%
  mutate(cumulative_count = cumsum(count)) %>%
  ungroup()

library(ggplot2)

ggplot(mutation_counts, aes(x = Sample, y = count, fill = ExonicFunc.refGene)) +
  geom_bar(stat = "identity", position = "stack") +
  theme_minimal() +
  labs(title = "Waterfall Plot of Mutation Types per Sample", x = "Sample ID", y = "Mutation Count") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_fill_brewer(palette = "Set3")

```


# 7.maftools
## maf data
```{r}
annovar.maf = annovarToMaf(annovar = "../03.Output/filtered_annovar_output.txt", 
                               Center = 'NA', 
                               refBuild = 'hg19', 
                               tsbCol = 'Tumor_Sample_Barcode', 
                               table = 'refGene',
                               sep = "\t") 

maf_data<- read.maf(maf =annovar.maf)

plotmafSummary(maf = maf_data,
               rmOutlier = TRUE,
               addStat = "median",
               dashboard = TRUE,
               titvRaw = FALSE)
```

## metadata
```{r}
design <- read.table("../01.Data/20211115-ClinMetaInfo-v1.2.tsv",header = T,row.names=1,sep = "\t")
design$patient <- rownames(design)

ITH_distance <- read.csv("../03.Output/ITH_distance.csv")
rownames(ITH_distance) <- ITH_distance$patient
ITH_distance <- ITH_distance%>%as.data.frame()
design$ITH <- ITH_distance[rownames(design),]$distance

metadata <- subset(annovar.maf,select=c("Tumor_Sample_Barcode")) %>%
  distinct(Tumor_Sample_Barcode,.keep_all = TRUE )

metadata$patient <- sapply(strsplit(as.character(metadata$Tumor_Sample_Barcode),"T"),"[",1)
metadata$patient <- sapply(strsplit(as.character(metadata$patient),"M"),"[",1)
metadata <- merge(metadata,design,by.x="patient",by.y="patient",all.x=T) 
metadata$patient <- factor(metadata$patient)
rownames(metadata) <- metadata$Tumor_Sample_Barcode
metadata <- metadata %>% as.data.frame()

drivergene <- read.table("../01.Data/driver_gene_4_Origins.txt",header = T,sep = ",") 
drivergene1 <- drivergene %>% distinct(Gene,.keep_all = TRUE) 

```

## oncoplot
```{R}

################### 
library(RColorBrewer)
qual_col_pals = brewer.pal.info[brewer.pal.info$category == 'qual',]
col_vector = unlist(mapply(brewer.pal, qual_col_pals$maxcolors, rownames(qual_col_pals)))
######################

oncoplot(maf = maf_data, top = 10,writeMatrix=T)

oncoplot(maf = maf_data,genes = drivergene1$Gene, fontSize = 1 ,showTumorSampleBarcodes = F,  #genes=top_genes, 
         draw_titv = TRUE,writeMatrix = TRUE)

mat <- read.table('../02.Code/onco_matrix.txt',sep = '\t') %>% as.data.frame()
head(mat)  

metadata_diff <- metadata[ metadata$Tumor_Sample_Barcode %in% colnames(mat),] %>% as.data.frame()

idx=as.character(maf_data@variants.per.sample$Tumor_Sample_Barcode[maf_data@variants.per.sample$Tumor_Sample_Barcode %in% colnames(mat)])
mat <- mat[,idx]
rownames(metadata_diff) <- metadata_diff$Tumor_Sample_Barcode
metadata_diff <- metadata_diff[ colnames(mat) ,]

mat <- mat[1:30,]

mat <- mat[rownames(rev(sort(apply(mat,1,function(x)sum(x != "")))) %>% as.data.frame()),]

col = c("Missense_Mutation" = "#a6bddb", 
         "Splice_Site" = "#beaed4", 
        "Nonsense_Mutation" = "#fdc086",
        "Nonstop_Mutation" ="#beaed4",
        "Frame_Shift_Del" = "#a6d854",
        "Frame_Shift_Ins" = "#e78ac3",
        "Multi_Hit" = "#fc8d62",
        "In_Frame_Del"= "#FFCD00",
        "In_Frame_Ins"= "#fbff7c"
        )

alter_fun = list(
  background = alter_graphic("rect",  width = 1, height = 1,fill = "white",col="#F2F2F2",lwd = 1),    #"#e9e9e9" "#F2F2F2"
  Missense_Mutation = alter_graphic("rect", width = 3, height = 1, fill = col["Missense_Mutation"],col="#F2F2F2"),
  Nonsense_Mutation = alter_graphic("rect", width = 3, height = 1,fill = col["Nonsense_Mutation"],col="#F2F2F2"),
  Nonstop_Mutation = alter_graphic("rect", width = 3, height = 1,fill = col["Nonstop_Mutation"],col="#F2F2F2"),
  Frame_Shift_Del = alter_graphic("rect",  width = 3, height = 1,fill = col["Frame_Shift_Del"],col="#F2F2F2"),
  Frame_Shift_Ins = alter_graphic("rect", width = 3, height = 1, fill = col["Frame_Shift_Ins"],col="#F2F2F2"),
  Splice_Site = alter_graphic("rect",  width = 3, height = 1,fill = col["Splice_Site"],col="#F2F2F2"),
  In_Frame_Del = alter_graphic("rect",  width = 3, height = 1,fill = col["In_Frame_Del"],col="#F2F2F2"),
  In_Frame_Ins = alter_graphic("rect", width = 3, height = 1, fill = col["In_Frame_Ins"],col="#F2F2F2"),
  Multi_Hit = alter_graphic("rect", width = 3, height = 1, fill = col["Multi_Hit"],col="#F2F2F2")
)

heatmap_legend_param = list(title = "Alternations type", 
                            grid_height = unit(4, "mm"),
                            grid_width = unit(4, "mm"),
                            legend_width = unit(1, "cm"),
                            at = c('Missense_Mutation', 'Nonsense_Mutation',"Nonstop_Mutation",
                                   'Splice_Site',
                                   'Frame_Shift_Del','Frame_Shift_Ins', 
                                   "In_Frame_Del","In_Frame_Ins",'Multi_Hit'), 
                            labels = c('Missense_Mutation', 'Nonsense_Mutation',"Nonstop_Mutation",
                                       'Splice_Site',
                                       'Frame_Shift_Del','Frame_Shift_Ins', 
                                       "In_Frame_Del","In_Frame_Ins",'Multi_Hit'))

ha<-HeatmapAnnotation( TMB=anno_barplot(maf_data@variants.per.sample[maf_data@variants.per.sample$Tumor_Sample_Barcode %in% colnames(mat),]$Variants ,border=FALSE, height = unit(0.5, "cm"),width  = unit(0.05, "cm"), gp = gpar(fill="grey2",fontsize = 0.5)),
                        show_annotation_name = TRUE,
                        annotation_name_gp = gpar(fontsize = 8))

pdf('../03.Output/oncoPrint_driver_gene.pdf',width = 9,height = 6)
oncoPrint(mat,
          alter_fun = alter_fun, 
          alter_fun_is_vectorized = FALSE,
          row_order = 1:nrow(mat),
          column_order = 1:ncol(mat),
          pct_gp = gpar(fontsize = 12), # 百分比的字体大小
          col = col,
          pct_side = "right",
          row_names_side = "left",
          row_names_gp = gpar(fontsize = 12),
          column_names_gp = gpar(fontsize = 12),
          # rect_gp = gpar(col = "grey", lwd = 1),
          # row_gap = unit(0.1, "mm"),
          show_heatmap_legend = F,
          top_annotation = ha,
          heatmap_legend_param = heatmap_legend_param,
          bottom_annotation = HeatmapAnnotation(
                                                patient=metadata_diff$patient,
                                                age= metadata_diff$age ,
                                                gender=metadata_diff$sex,
                                                ITH=metadata_diff$ITH,
                                                annotation_name_gp = gpar(fontsize = 12),
                                                simple_anno_size = unit(0.3, "cm"),
                                                show_legend = c( FALSE,FALSE, FALSE, FALSE),
                                                # show_legend = c( FALSE,TRUE, TRUE, TRUE),
                                               col = list(
  patient = structure(col_vector[c(1:30, 46:71)], names = levels(factor(metadata_diff$patient))),
  gender = c("Male" = "#c6eae4", "Female" = "#80cdc1"),
  age = colorRamp2(c(10, 80), c("white", "#8DA0CB")),
  ITH = colorRamp2(c(0.1, 1), c("white", "#9E8A7A"))
      )
  )
  )#, heatmap_legend_param = heatmap_legend_param)

dev.off()

```


# 8.patients level
```{r}
mut = maf_data@data[, c("Hugo_Symbol",
                           "Chromosome",
                           "Start_Position",
                           "End_Position",
                           "ExonicFunc.refGene",
                           "Tumor_Sample_Barcode")]

mut$patient <- sapply(strsplit(as.character(mut$Tumor_Sample_Barcode),"T"),"[",1)
mut$patient <- sapply(strsplit(as.character(mut$patient),"M"),"[",1)
mut$tumorsite <- sapply(strsplit(as.character(mut$Tumor_Sample_Barcode),"T"),"[",2)
mut$tumorsite <- ifelse(!is.na(mut$tumorsite),paste0("T",mut$tumorsite),"M")
mut$mutation_id <- paste0(mut$Chromosome,"_",mut$Start_Position,"_",mut$End_Position)

mut$patient <- factor(mut$patient)

mut1 <- mut[mut$Hugo_Symbol %in% drivergene1$Gene,]

a <- sort(table(mut1$Hugo_Symbol)) %>% as.data.frame()

```

## metadata
```{r}

metadata <- read.table("../01.Data/metadata_WES.tsv",header = T,row.names=1,sep = "\t")

metadata$patient1 <- metadata$patient
metadata$patient1 <- gsub("HCC","S",metadata$patient1)

```

##  mutationid - label
```{r}
dir.create("../03.Output/patient1")

cust_label <- read.csv("../01.Data/cust_label.csv",row.names = 1)
colnames(cust_label) <- "name_change"
cust_label$name <- gsub("S","HCC",rownames(cust_label))

mutation_id <- list()
percentage_mutationid <- list()
calc_ht_size = function(ht, unit = "inch") {
  pdf(NULL)
  ht = draw(ht)
  w = ComplexHeatmap:::width(ht)
  w = convertX(w, unit, valueOnly = TRUE)
  h = ComplexHeatmap:::height(ht)
  h = convertY(h, unit, valueOnly = TRUE)
  dev.off()
  c(w, h)
}
for (i in 1:length(unique(mut$patient))){
  for (j in levels(unique(mut$patient))[i]){
    print(j)
    mat0 <- unique(mut[mut$patient==j,c("tumorsite","mutation_id","Hugo_Symbol","ExonicFunc.refGene")]) # 
    mat0$ExonicFunc.refGene <- ifelse(mat0$ExonicFunc.refGene=="-","splicing",mat0$ExonicFunc.refGene)

    mat=mat0 %>% data.frame
    mat <- mat[,-which(colnames(mat) %in% c("Hugo_Symbol"))] #,"ExonicFunc.refGene"
    mat=spread(mat,tumorsite,ExonicFunc.refGene,fill=0) %>% tibble::column_to_rownames(var="mutation_id")
    mat[,1:ncol(mat)] <- lapply(mat[,1:ncol(mat)],as.factor)
    mutation_id[[i]] <- mat
    names(mutation_id)[[i]] <-j
    
    mat_SNV <- apply(mat,2,function(x){
      sum(!x=="0")
    }) %>% as.matrix()
    
    mat_spread=mat %>% as.data.frame
    mat_spread$count <- apply(mat_spread,1,function(x){
      sum(!x=="0")
    })
    mat_spread$label <- ifelse(mat_spread$count==ncol(mat),"trunk",
                               ifelse(mat_spread$count==1,"private","brunch"))
    mat_spread$label <- factor(mat_spread$label,levels = c("trunk","brunch","private"))
    percentage_trunk <- round((sum(mat_spread$count==ncol(mat))/nrow(mat))*100,2)
    percentage_brunch <- round(sum(mat_spread$count<ncol(mat) & mat_spread$count>1)/nrow(mat)*100,2)
    percentage_private <- round((sum(mat_spread$count==1)/nrow(mat))*100,2)
    percentage_mutationid[[i]] <- data.frame(Tumorsites=ncol(mat),SNVs=nrow(mat),
                                             Trunk_mutations=percentage_trunk,
                                             Branch_mutations=percentage_brunch,
                                             Private_mutations=percentage_private)
    percentage_mutationid[[i]] <- cbind(percentage_mutationid[[i]],t(mat_SNV))
    names(percentage_mutationid)[[i]] <- j
    for (i in 1:ncol(mat)){
      if (i==3){
        dat=mat_spread[order(mat_spread$count,mat_spread[,1],mat_spread[,2],mat_spread[,3],decreasing = TRUE),] 
      } else
        if(i==4){
          dat=mat_spread[order(mat_spread$count,mat_spread[,1],mat_spread[,2],mat_spread[,3],mat_spread[,4],decreasing = TRUE),]
        } else
          if(i==5){
            dat=mat_spread[order(mat_spread$count,mat_spread[,1],mat_spread[,2],mat_spread[,3],mat_spread[,4],mat_spread[,5],decreasing = TRUE),]
          } else
            if(i==6){
              dat=mat_spread[order(mat_spread$count,mat_spread[,1],mat_spread[,2],mat_spread[,3],mat_spread[,4],mat_spread[,5],mat_spread[,6],decreasing = TRUE),]
            } else
              if(i==7){
                dat=mat_spread[order(mat_spread$count,mat_spread[,1],mat_spread[,2],mat_spread[,3],mat_spread[,4],mat_spread[,5],mat_spread[,6],mat_spread[,7],decreasing = TRUE),]
              }
    }
    dat1 <- merge(dat,mat0[,c("mutation_id","Hugo_Symbol")],by.x="row.names",by.y="mutation_id",sort=F)
    dat1 <- dat1[!duplicated(dat1),] %>% as.data.frame 
    rownames(dat1) <- dat1$Row.names
    dat1 <- dat1[,-1]
    dat1$drivergene <- ifelse(dat1$Hugo_Symbol %in% drivergene1$Gene,"YES","NO")
    write.csv(dat1,file=paste0("../03.Output/patient1/",j,"_dat.csv"))
    
    drivermark <- which( dat1$Hugo_Symbol %in% drivergene1$Gene)
    labs <- dat1$Hugo_Symbol[drivermark] 

    ha <-  rowAnnotation(type=dat1$label, simple_anno_size = unit(2, "mm"), #width = unit(2, "mm"),
                         col=list(type=c("trunk"="#a64942","brunch"="#f29c2b","private"="#8bc24c")), 
                         show_legend = F,show_annotation_name = FALSE, #"#7B3B8C"
                         foo = anno_mark(at = drivermark,
                                         labels = labs,
                                         labels_gp = gpar(fontsize = 8))
    )
    colors <- structure(c("#E6E6E6","#6a60a9","#b689b0","#d5a4cf","#f9bcdd","#dd7777","#ffb6b9","#ffe5e1","#fae3d9"),
                        names = c("0", "nonsynonymous SNV","frameshift insertion","frameshift deletion","nonframeshift insertion","nonframeshift deletion","stopgain","stoploss","splicing")
                        #c("#145593","firebrick3","#4f0804","#9c331e","#bb6b4e","#b38766","#ddbb93","#e4ccb3")) 
    ) #"#fffff5"

    p <- ComplexHeatmap::Heatmap(as.matrix(dat1[,1:ncol(mat)]),col = colors, #border = T,
                                 #name = paste0(j," (",percentage,"%)"),
                                 #rect_gp = gpar(col = "#F5F7FA", lwd = 0.001),
                                 cluster_rows = F,
                                 cluster_columns = F,
                                 right_annotation = ha,
                                 show_row_names = F,
                                 show_heatmap_legend = F,
                                 # column_title=paste0(j," (",percentage_trunk,"%)"),
                                 column_title=paste0(cust_label[cust_label$name==j,]$name_change),
                                 column_title_gp=gpar(fontsize = 12), #18
                                 column_names_rot = 0,
                                 column_names_centered = TRUE,
                                 #column_km = 4,
                                 #column_split = ncol(dat[,1:ncol(mat)]),
                                 #row_names_gp = gpar(fontsize = 10),
                                 column_names_gp = gpar(fontsize = 12), #20
                                 width=unit(5, "mm")*ncol(dat1[,1:ncol(mat)]),
                                 height = unit(0.5, "mm")*nrow(dat1[,1:ncol(mat)])
                                 )
    print(p)
    size = calc_ht_size(p)
    width = size[1]+0.1
    height = size[2]
    
    pdf(paste0("../03.Output/patient/",j,"_",paste0(cust_label[cust_label$name==j,]$name_change),"_withlabel.pdf"), height = height , width =width)
    print(p)
    dev.off()
  }
}

```

##  mutationid - no label
```{r}

mutation_id <- list()
percentage_mutationid <- list()
calc_ht_size = function(ht, unit = "inch") {
  pdf(NULL)
  ht = draw(ht)
  w = ComplexHeatmap:::width(ht)
  w = convertX(w, unit, valueOnly = TRUE)
  h = ComplexHeatmap:::height(ht)
  h = convertY(h, unit, valueOnly = TRUE)
  dev.off()
  c(w, h)
}
for (i in 1:length(unique(mut$patient))){
  for (j in levels(unique(mut$patient))[i]){
    print(j)
    mat0 <- unique(mut[mut$patient==j,c("tumorsite","mutation_id","Hugo_Symbol","ExonicFunc.refGene")]) 
    mat0$ExonicFunc.refGene <- ifelse(mat0$ExonicFunc.refGene=="-","splicing",mat0$ExonicFunc.refGene)

    mat=mat0 %>% data.frame
    mat <- mat[,-which(colnames(mat) %in% c("Hugo_Symbol"))] #,"ExonicFunc.refGene"
    mat=spread(mat,tumorsite,ExonicFunc.refGene,fill=0) %>% tibble::column_to_rownames(var="mutation_id")
    mat[,1:ncol(mat)] <- lapply(mat[,1:ncol(mat)],as.factor)
    mutation_id[[i]] <- mat
    names(mutation_id)[[i]] <-j
    
    mat_SNV <- apply(mat,2,function(x){
      sum(!x=="0")
    }) %>% as.matrix()
    
    mat_spread=mat %>% as.data.frame
    mat_spread$count <- apply(mat_spread,1,function(x){
      sum(!x=="0")
    })
    mat_spread$label <- ifelse(mat_spread$count==ncol(mat),"trunk",
                               ifelse(mat_spread$count==1,"private","brunch"))
    mat_spread$label <- factor(mat_spread$label,levels = c("trunk","brunch","private"))
    percentage_trunk <- round((sum(mat_spread$count==ncol(mat))/nrow(mat))*100,2)
    percentage_brunch <- round(sum(mat_spread$count<ncol(mat) & mat_spread$count>1)/nrow(mat)*100,2)
    percentage_private <- round((sum(mat_spread$count==1)/nrow(mat))*100,2)
    percentage_mutationid[[i]] <- data.frame(Tumorsites=ncol(mat),SNVs=nrow(mat),
                                             Trunk_mutations=percentage_trunk,
                                             Branch_mutations=percentage_brunch,
                                             Private_mutations=percentage_private)
    percentage_mutationid[[i]] <- cbind(percentage_mutationid[[i]],t(mat_SNV))
    names(percentage_mutationid)[[i]] <- j
    for (i in 1:ncol(mat)){
      if (i==3){
        dat=mat_spread[order(mat_spread$count,mat_spread[,1],mat_spread[,2],mat_spread[,3],decreasing = TRUE),] 
      } else
        if(i==4){
          dat=mat_spread[order(mat_spread$count,mat_spread[,1],mat_spread[,2],mat_spread[,3],mat_spread[,4],decreasing = TRUE),]
        } else
          if(i==5){
            dat=mat_spread[order(mat_spread$count,mat_spread[,1],mat_spread[,2],mat_spread[,3],mat_spread[,4],mat_spread[,5],decreasing = TRUE),]
          } else
            if(i==6){
              dat=mat_spread[order(mat_spread$count,mat_spread[,1],mat_spread[,2],mat_spread[,3],mat_spread[,4],mat_spread[,5],mat_spread[,6],decreasing = TRUE),]
            } else
              if(i==7){
                dat=mat_spread[order(mat_spread$count,mat_spread[,1],mat_spread[,2],mat_spread[,3],mat_spread[,4],mat_spread[,5],mat_spread[,6],mat_spread[,7],decreasing = TRUE),]
              }
    }
    dat1 <- merge(dat,mat0[,c("mutation_id","Hugo_Symbol")],by.x="row.names",by.y="mutation_id",sort=F)
    dat1 <- dat1[!duplicated(dat1),] %>% as.data.frame 
    rownames(dat1) <- dat1$Row.names
    dat1 <- dat1[,-1]
    
    drivermark <- which( dat1$Hugo_Symbol %in% drivergene1$Gene)
    labs <- dat1$Hugo_Symbol[drivermark] 
    
    ha <-  rowAnnotation(type=dat1$label, simple_anno_size = unit(2, "mm"), #width = unit(2, "mm"),
                         col=list(type=c("trunk"="#a64942","brunch"="#f29c2b","private"="#8bc24c")), 
                         show_legend = F,show_annotation_name = FALSE #"#7B3B8C"
                         # foo = anno_mark(at = drivermark,
                         #                 labels = labs,
                         #                 labels_gp = gpar(fontsize = 15))
                         )
    colors <- structure(c("#E6E6E6","#6a60a9","#b689b0","#d5a4cf","#f9bcdd","#dd7777","#ffb6b9","#ffe5e1","#fae3d9"),
                        names = c("0", "nonsynonymous SNV","frameshift insertion","frameshift deletion","nonframeshift insertion","nonframeshift deletion","stopgain","stoploss","splicing")
                        #c("#145593","firebrick3","#4f0804","#9c331e","#bb6b4e","#b38766","#ddbb93","#e4ccb3")) 
                        ) #"#fffff5"

    p <- ComplexHeatmap::Heatmap(as.matrix(dat1[,1:ncol(mat)]),col = colors, #border = T,
                                  #name = paste0(j," (",percentage,"%)"),
                                  #rect_gp = gpar(col = "#F5F7FA", lwd = 0.001),
                                  cluster_rows = F,
                                  cluster_columns = F,
                                  right_annotation = ha,
                                  show_row_names = F,
                                  show_heatmap_legend = F,
                                  # column_title=paste0(j," (",percentage_trunk,"%)"),
                                  column_title=paste0(cust_label[cust_label$name==j,]$name_change),
                                  column_title_gp=gpar(fontsize = 12), #18
                                  column_names_rot = 0,
                                  column_names_centered = TRUE,
                                  #column_km = 4,
                                  #column_split = ncol(dat[,1:ncol(mat)]),
                                  #row_names_gp = gpar(fontsize = 10),
                                  column_names_gp = gpar(fontsize = 12), #20
    width=unit(5, "mm")*ncol(dat1[,1:ncol(mat)]),
    height = unit(0.1, "mm")*nrow(dat1[,1:ncol(mat)]))
    print(p)
    size = calc_ht_size(p)
    width = size[1]+0.1
    height = size[2]
    
    pdf(paste0("../03.Output/patient/",j,"_",paste0(cust_label[cust_label$name==j,]$name_change),"_1.pdf"), height = height , width =width)
    print(p)
    dev.off()
  }
}

```

## bar plot
```{r}
percentage_mutationid_1 <- do.call(rbind, lapply(lapply(percentage_mutationid, unlist), `length<-`, max(lengths(percentage_mutationid))))%>% as.data.frame

percentage_mutationid_1 <- percentage_mutationid_1[order(percentage_mutationid_1$Trunk_mutations,percentage_mutationid_1$Branch_mutations,percentage_mutationid_1$Private_mutations),] %>% 
  tibble::rownames_to_column(var = "Patients")
percentage_mutationid_1 <- merge(percentage_mutationid_1,cust_label,by.x="Patients",by.y="name", all.x = T)
write.csv(percentage_mutationid_1,"../03.Output/percentage_mutationid.csv")

percentage_mutationid_1 <- percentage_mutationid_1 %>%
  arrange(desc(Trunk_mutations),desc(Branch_mutations))

percentage_mutationid_1_long <- reshape2::melt(percentage_mutationid_1[,-c(2:3,7:14)],
                                   id.vars="Patients", 
                                   variable.name = "Type",
                                   value.name = "Percentage")


percentage_mutationid_1_long$name_change <- cust_label[match(percentage_mutationid_1_long$Patients,cust_label$name),]$name_change
percentage_mutationid_1_long <- percentage_mutationid_1_long[!is.na(percentage_mutationid_1_long$name_change),]

percentage_mutationid_1_long <- percentage_mutationid_1_long %>%
  mutate(Type=factor(Type,levels=c("Trunk_mutations","Branch_mutations","Private_mutations"))) %>% 
  arrange(desc(Type == "Trunk_mutations"),desc(Percentage)) %>%
   mutate(name_change = factor(name_change, levels = percentage_mutationid_1$name_change))

p_percentage <-   ggplot(data=percentage_mutationid_1_long, mapping=aes(x=name_change,y=Percentage,fill=Type))+ #reorder(Patients,Type)
  geom_bar(stat='identity',position='fill')+
  labs(x = '',y = 'Percentage(%)') +
  scale_y_continuous(labels = scales::percent,expand = c(0,0)) +
  scale_fill_manual(values=c("#a64942","#f29c2b","#8bc24c"),
                    name="Type",
                    breaks=c("Trunk_mutations","Branch_mutations","Private_mutations"),
                    labels=c("Trunk mutations","Branch mutations","Private mutations"))+
  theme(panel.background=element_rect(fill="white"),
        panel.border = element_rect(fill=NA,size=0.1),
        axis.title.y=element_text(size=14), 
        axis.title.x=element_blank(),
        axis.line=element_line(color='black', size=0.1),      
        axis.text.x=element_text(color='black', size=8,angle = 90, hjust = 1,vjust = 0.5), 
        axis.text.y=element_text(color='black', size=12), 
        axis.ticks=element_line(color='black', size=0.1),
        axis.ticks.length=unit(0.01, "cm"),
        legend.position="top",
        legend.key=element_blank(),
        legend.key.height = unit(5, "pt"),
        legend.key.width  = unit(5, "pt"),
        legend.title=element_text(color='black', size=14),
        legend.text=element_text(color='black', size=12));p_percentage

ggsave("../03.Output/percentage_mutationid.pdf",p_percentage,width=6,height=4)

```


## ITH - by jaccard distance
```{r}
ITH_distance <- data.frame(patient=levels(unique(mut$patient)),distance=rep("NA",length(unique(mut$patient))))

for (i in 1:length(unique(mut$patient))){
  for (j in levels(unique(mut$patient))[i]){
    print(j)
    mat <- unique(mut[mut$patient==j,c("Tumor_Sample_Barcode","mutation_id")]) #Hugo_Symbol #mutation_id
    mat$tmp=1

    mat=spread(mat,Tumor_Sample_Barcode,tmp,fill=0)
    mat=as.data.frame(mat)
    rownames(mat)=mat$mutation_id
    mat=mat[,-1]
    gene.dist <- vegan::vegdist(t(mat), method = 'jaccard',na.rm = TRUE) #robust.aitchison #euclidean
    ITH_distance[ITH_distance$patient==j,]$distance <- mean(as.vector(gene.dist))
  }
}

ITH_distance$distance <- as.numeric(ITH_distance$distance)
ITH_distance <- ITH_distance[order(ITH_distance$distance),]
ITH_distance$patient <- as_factor(ITH_distance$patient)
median(ITH_distance$distance)
range(ITH_distance$distance)
write.csv(ITH_distance,"../03.Output/ITH_distance.csv",row.names = F)

p_ITH <-  ggplot(data=ITH_distance, mapping=aes(x=patient,y=distance,group=1))+ #reorder(Patients,Type)
  geom_point(color="red")+
  geom_line()+
  labs(x = '',y = 'Inter-tumor heterogeneity (jaccard distance)') +
  # scale_fill_manual(values=c("#7B3B8C","#f29c2b","#8bc24c"))+
  theme(panel.grid=element_blank(), 
        panel.background=element_rect(color='black', fill='transparent'),
        axis.text.x = element_text(angle = 90, hjust = 1,vjust = 0.5),
        axis.title =element_text(size = 14),
        axis.text =element_text(size = 12, color = 'black'));p_ITH

ggsave("../03.Output/ITH.pdf",p_ITH,width=9,height=5)

```

## genomic - by euclidean distance
```{r}
####计算 jaccard distance
euclidean_distance <- list()
for (i in 1:length(unique(mut$patient))){
  for (j in levels(unique(mut$patient))[i]){
    print(j)
    mat <- unique(mut[mut$patient==j,c("Tumor_Sample_Barcode","mutation_id")]) #Hugo_Symbol #mutation_id
    mat$tmp=1

    mat=spread(mat,Tumor_Sample_Barcode,tmp,fill=0)
    mat=as.data.frame(mat)
    rownames(mat)=mat$mutation_id
    mat=mat[,-1]
    d1 <- dist(t(mat), method = "euclidean", diag = FALSE, upper = FALSE, p = 2 ) %>% as.matrix %>% #as.dist
      as.data.frame %>%
      tibble::rownames_to_column(var="sampleid")
    d1_long <- reshape2::melt(d1,
                              id.vars = "sampleid",
                              variable.name = "paired_sampleid",
                              value.name = "distance")
    d2 <- t(combn(d1$sampleid,2)) %>% as.data.frame 
    colnames(d2) <- c("sampleid","paired_sampleid")
    d2$distance <- apply(d2,1,function(ar){
      d1_long[d1_long$sampleid==ar[1] &d1_long$paired_sampleid==ar[2],]$distance
    })
    euclidean_distance[[i]] <- d2
    names(euclidean_distance)[[i]] <- j
  }
}

distance <- do.call(rbind,euclidean_distance)

distance1 <- merge(distance,metadata[,c("Tumor_Sample_Barcode","patient","nodule","nodule1")],by.x="sampleid",by.y="Tumor_Sample_Barcode")
distance1$paired <- paste0(distance1$sampleid,"_",distance1$paired_sampleid)
distance1$nodule1 <- factor(distance1$nodule1,levels = c("MO","MIX_MO","MIX_IM","IM"))
distance1 <- distance1[order(distance1$nodule1),]

distance1$patient <- factor(distance1$patient) 

distance1$paired <- as_factor(distance1$paired) #,ordered = FALSE

kruskal.test(distance ~ nodule1, data = distance1)
dunnTest(distance ~ nodule1, data = distance1)

distance2 <- distance1[distance1$nodule %in% c("MO","IM"),]
wilcox.test(distance ~ nodule, data = distance2,
            exact = FALSE)

distance3 <- distance1 %>%
  group_by(nodule1) %>%
  summarise(y=mean(distance))

p_distance <- ggplot(data=distance1, mapping=aes(x=paired,y=distance))+ #reorder(Patients,Type)
  geom_bar(stat='identity')+
  labs(x = '',y = 'Distance') +
  #scale_fill_manual(values=c("#7B3B8C","#f29c2b","#8bc24c"))+
  theme(axis.text.x = element_text(angle = 90, hjust = 1,vjust = 0.5),
        axis.title =element_text(size = 10),
        axis.text =element_text(size = 5, color = 'black')) 
p_distance
ggsave("../03.Output/distance1.pdf",p_distance,width=9,height=5)

p_distance <- ggplot(data=distance1, mapping=aes(x=nodule1,y=distance,fill=nodule1))+ #reorder(Patients,Type)
  geom_boxplot(width=0.5,outlier.size = 0.1,linewidth=0.1)+
  labs(x = '',y = 'Within group genomic distance') +
  geom_signif(comparisons = list(c("MO","MIX_MO"),c("MO","MIX_IM"), c("MO","IM"),
                                 c("IM","MIX_MO"),c("IM","MIX_IM")),  
              map_signif_level = TRUE,
              size = 0.2,
              textsize = 4,
              test = "wilcox.test",
              y_position = c(56,60,64,68,74)
              )+ #tip_length = c(0.05,0.05))+
  scale_fill_manual(values = c("#4393c3","#92c5de","#f4a582","#d6604d"))+
  guides(fill="none")+
  # guides(fill=guide_legend(title="Nodule type")) +
  theme(panel.background=element_rect(fill="white"),
        panel.border = element_rect(fill=NA,size=0.2),
        # text=element_text(color='black', size=8),
        axis.title.x=element_blank(), 
        axis.title.y=element_text(size=14), 
        axis.line=element_line(size=0.1),      
        axis.text.x=element_text(color='black', size=12,hjust = 0.5,vjust = 0.5), #angle = 90, 
        axis.text.y=element_text(color='black', size=12), 
        axis.ticks=element_line(color='black', size=0.1),
        axis.ticks.length=unit(0.01, "cm"),
        legend.key=element_blank(),
        legend.key.height = unit(5, "pt"),
        legend.key.width  = unit(5, "pt"),
        legend.title=element_text(color='black', size=14),
        legend.text=element_text(color='black', size=12),
        # title = element_text(color='black', size=13)
  );p_distance

ggsave("../03.Output/distance2.pdf",p_distance,width=3.5,height=2.8)


```

## TMB
```{r}

total_bases <- 60 * 10^6 

tmb_data <- tmb(maf = maf_data, captureSize = total_bases)

tmb_df <- data.frame(
  Tumor_Sample_Barcode = tmb_data$Tumor_Sample_Barcode,
  TMB = tmb_data$total_perMB
)

ggplot(tmb_df, aes(x = Tumor_Sample_Barcode, y = TMB)) +
  geom_bar(stat = "identity", fill = "steelblue") +
  theme_minimal() +
  labs(title = "Tumor Mutation Burden (TMB) per Sample",
       x = "Sample",
       y = "TMB (mutations per Mb)") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

TMB <- merge(tmb_df,metadata[,c("Tumor_Sample_Barcode","patient","nodule","nodule1")],by.x="Tumor_Sample_Barcode",by.y="Tumor_Sample_Barcode")
TMB$nodule1 <- factor(TMB$nodule1,levels = c("MO","MIX_MO","MIX_IM","IM"))
TMB <- TMB[order(TMB$nodule1),]

kruskal.test(TMB ~ nodule1, data = TMB)
dunnTest(TMB ~ nodule1, data = TMB)

TMB1 <- TMB[TMB$nodule %in% c("MO","IM"),]
wilcox.test(TMB ~ nodule, data = TMB1,
            exact = T)

TMB2 <- TMB %>%
  group_by(nodule1) %>%
  summarise(y=mean(TMB))

p_TMB <- ggplot(data=TMB, mapping=aes(x=nodule1,y=TMB,fill=nodule1))+ #reorder(Patients,Type)
  geom_boxplot(width=0.5,outlier.size = 0.1,linewidth=0.1)+
  labs(x = '',y = 'Tumor Mutation Burden') +
  geom_signif(comparisons = list(c("MO","MIX_MO"),c("MO","MIX_IM")),   #c("MIX_MO","MIX_IM"),c("MO","MIX_MO"),c("IM","MIX_IM")), 无意义
              map_signif_level = TRUE,
              size = 0.2,
              textsize = 4,
              test = "wilcox.test",
              y_position = c(2.8e-05,3.00e-05)
  )+ #tip_length = c(0.05,0.05))+
  scale_fill_manual(values = c("#4393c3","#92c5de","#f4a582","#d6604d"))+
  guides(fill="none")+
  # guides(fill=guide_legend(title="Nodule type")) +
  theme(panel.background=element_rect(fill="white"),
        panel.border = element_rect(fill=NA,size=0.2),
        # text=element_text(color='black', size=8),
        axis.title.x=element_blank(), 
        axis.title.y=element_text(size=14), 
        axis.line=element_line(size=0.1),      
        axis.text.x=element_text(color='black', size=12,hjust = 0.5,vjust = 0.5), #angle = 90, 
        axis.text.y=element_text(color='black', size=12), 
        axis.ticks=element_line(color='black', size=0.1),
        axis.ticks.length=unit(0.01, "cm"),
        legend.key=element_blank(),
        legend.key.height = unit(5, "pt"),
        legend.key.width  = unit(5, "pt"),
        legend.title=element_text(color='black', size=14),
        legend.text=element_text(color='black', size=12),
        # title = element_text(color='black', size=13)
  );p_TMB

ggsave("../03.Output/TMB.pdf",p_TMB,width=3.7,height=2.8)

```

# 9.mutation signature
```{r}
library(deconstructSigs)
library('BSgenome.Hsapiens.UCSC.hg19')
library(pheatmap)

maf <- maf_data@data

colnames(maf)

sample.mut.ref<-data.frame(Sample=maf[,6],
                           chr =maf[,1], 
                           pos = maf[,2], 
                           ref = maf[,4],
                           alt = maf[,5]) #57037 43134
colnames(sample.mut.ref) <- c("Sample","chr","pos","ref","alt")

data(signatures.cosmic)

sigs.input<-mut.to.sigs.input(mut.ref = sample.mut.ref,
                                sample.id = "Sample",
                                chr = "chr",
                                pos = "pos",
                                ref = "ref",
                                alt = "alt",
                                bsg =BSgenome.Hsapiens.UCSC.hg19)

df = data.frame()
for(i in rownames(sigs.input)){
  print(i)
  sigs.output <-whichSignatures(tumor.ref=sigs.input,
                                signatures.ref =signatures.cosmic, 
                                sample.id =i,
                                contexts.needed =TRUE)
  df =rbind(df,sigs.output$weights)
}

df1 = df[,apply(df,2,function(x){sum(x>0)})>10]

pheatmap(df1,
         scale = "row",
         cellwidth = 5, 
         cellheight = 3,
         cluster_cols = F,
         cluster_rows = T,
         fontsize =5)

```

## pheatmap
```{r}

df2 <- t(df1)
rownames(metadata) <- metadata$Tumor_Sample_Barcode
df2 <- df2[,rownames(metadata)]


RowV <- as.matrix(df1) %>%
  factoextra::get_dist(method = 'euclidean') %>% # "euclidean", "maximum", "manhattan", "canberra", "binary", "minkowski", "pearson", "spearman" or "kendall"
  hclust(method = 'ward.D2') %>% rev %>% # "ward.D", "ward.D2", "single", "complete", "average" (= UPGMA), "mcquitty" (= WPGMA), "median" (= WPGMC) or "centroid" (= UPGMC).
  as.dendrogram
# save(RowV, file = 'Result/2.scMethod-allOTU-GI-heatmap.RowV.Rds')

ordering_row <- seq(dim(df1)[1]); names(ordering_row) <- rownames(df1)
ordering_row <- ordering_row[order.dendrogram(RowV)]

######################
p <-ComplexHeatmap::Heatmap(mat=df2,col = colorRampPalette(c("white","red"))(2),border=T,
                        #name = paste0(p," (",percentage,"%)"),
                        # rect_gp = gpar(col = "white", lwd = 0.01),
                        border_gp = gpar(col = "black",lwd=0.1),
                        cluster_rows = F,
                        cluster_columns = F,
                        show_column_dend = F,
                        show_heatmap_legend = FALSE,
                        # right_annotation = rowAnnotation(patient=metadata$patient,
                        #                                  col=list(Patient=structure(names=c(levels(unique(metadata$patient))),c(col_vector[1:58])))),
                        top_annotation = HeatmapAnnotation(Age=metadata$age,Gender=metadata$sex,Smoking=metadata$Smoking,Alcohol=metadata$Alcohol,
                                                           # TBIL=metadata$TBil,DBIL=metadata$DBil,ALT=metadata$ALT,AST=metadata$AST,
                                                           # HBsAg=metadata$HBsAg,HCV=metadata$HepC,
                                                           BCLC=metadata$BCLC,Patients=metadata$patient,Nodule=metadata$nodule1,
                                                           annotation_name_gp = gpar(fontsize = 12),
                                                           simple_anno_size = unit(0.3, "cm"),
                                                           show_annotation_name = T,annotation_name_side ="left", show_legend = c(TRUE,TRUE,TRUE,TRUE,TRUE,FALSE,TRUE),
                                                           col=list(Patients=structure(names=c(levels(unique(metadata$patient))),c(col_vector[1:56])),
                                                                    Nodule = structure(names = c("MO","MIX_MO","MIX_IM","IM"),
                                                                                      c("#ea8943", "#F7B733","#8db842","#4a7246")),
                                                                    Age = colorRamp2(c(0, 80), c("white", "#2D4272")),
                                                                    Gender = structure(names = c("Female", "Male"), c("#0E9B4D", "#E835BC")),
                                                                    Smoking=structure(names = c("smoking", "no_smoking","no_report"), c("#7e0000", "#fcd39e","white")),
                                                                    Alcohol=structure(names = c("alcohol", "no-alcohol","no_report"), c("#806751", "#cca233","white")),
                                                                    # BMI = structure(names = c("Underweight", "Normal","Overweight","Obese"),
                                                                    #                 c("#7e0000","#fcd39e", "#fb8d58", "#d62f1f" )),
                                                                    # HBsAg = structure(names = c("positive", "negative"), c("#cca233", "white")),
                                                                    # HCV = structure(names = c("positive", "negative"), c("#806751", "white")),
                                                                    BCLC=structure(names = c("1", "2","3","4","5","no_report"), 
                                                                                   c("#2E94B9", "#FFFDC0", "#F0B775", "#D25565","#E8222D","white")))),
                        show_row_names = T,
                        show_column_names = F,
                        # heatmap_legend_param = list(legend_height = unit(1, "cm"),legend_weight = unit(1, "cm"),),
                        # show_heatmap_legend = T,
                        #column_title=paste0(p," (",percentage,"%)"),
                        #column_title_gp=gpar(fontsize = 18),
                        column_names_rot = 90,
                        column_names_centered = F,
                        row_names_gp = gpar(fontsize = 12),
                        column_names_gp = gpar(fontsize = 12))
                        # width=unit(0.2, "cm")*ncol(df2),
                        # height = unit(0.5, "cm")*nrow(df2))
p

pdf("../03.Output/signature_nodule.pdf",width = 10,height = 7)
p
dev.off()

```


## nodule
```{r}
df3 <- reshape2::melt(df2,
                      id.vars="Signature", 
                      variable.name = "sampleid",
                      value.name = "propotion")
colnames(df3) <-c("Signature","sampleid","propotion")
df3$nodule <- metadata[match(df3$sampleid,metadata$Tumor_Sample_Barcode),]$nodule1
df3$nodule <- factor(df3$nodule,levels = c("MO","MIX_MO","MIX_IM","IM"))

library(ggsignif)
p_diff <-   ggplot(data=df3, mapping=aes(x=nodule,y=propotion,fill=nodule))+ #reorder(Patients,Type)
  geom_boxplot(width=0.5,outlier.size = 0.3,linewidth=0.2)+
  labs(x = '',y = '') + #log2(count+1)
  facet_wrap(~Signature,scales = "free_y",ncol=6)+
  # scale_fill_manual(values=c("#7B3B8C","#f29c2b","#8bc24c","#F67280"))+
  geom_signif(comparisons = list(c("MO", "IM")),
              map_signif_level = TRUE,
              size = 0.2,
              textsize = 2,
              test = "wilcox.test")+
  guides(fill="none")+
  # guides(fill=guide_legend(title="Nodule type")) +
  #scale_y_continuous(limits = c(0,16))+
  theme(panel.background=element_blank(),
        panel.border = element_rect(fill=NA,size=0.1),
        # panel.spacing.x = unit(0.05,"cm"),
        axis.text.x=element_text(color='black', size=6), 
        axis.text.y=element_text(color='black', size=8), 
        axis.ticks.length =unit(0.05, "cm"),
        axis.ticks = element_line(color='black', size=0.1),
        axis.title.y = element_text(color='black', size=14), 
        legend.position="top",
        legend.title=element_text(color='black', size=14),
        legend.text=element_text(color='black', size=12,face="italic"),
        legend.key=element_blank(),
        legend.key.height = unit(8, "pt"),
        legend.key.width  = unit(8, "pt"),
        strip.text = element_text(size=12), #hcc
        strip.background = element_blank()) 

  # theme(panel.grid.major = element_blank(),
  #       panel.grid.minor = element_blank(),
  #       panel.background=element_blank(),
  #       panel.border = element_rect(color = "black", linewidth = 1, fill = NA),
  #       #axis.title.x=element_text(size=12,face = "bold"),
  #       axis.title.y=element_text(size=12,face = "bold"),
  #       axis.text = element_text(size = 10,face = "bold"),
  #       axis.text.x = element_text(angle = 90, hjust = 1,vjust = 0.5),
  #       axis.ticks.x = element_blank(),
  #       strip.text = element_text(face = "bold",size=12),
  #       legend.text = element_text(size = 11),
  #       legend.key.size = unit(12, "pt"))
p_diff

ggsave("../03.Output/signature_nodule_difference.pdf",width = 10,height = 5)

```
